### PR TITLE
Jobgraph cli

### PR DESCRIPTION
**Pull request by KKRainbow**

### CODE REVIEW

File: cli/entry.go [hunk 1]
```diff
@@ -0,0 +1,360 @@
+package cli
+
+import (
+	"fmt"
```
Comment (Lcch) [inline cli/entry.go]: import style

File: master/master.go [hunk 2]
```diff
@@ -91,19 +92,14 @@ func (m *Master) CheckHeartbeatForEachWorker(workerID int64, heartbeat chan int)
 		case <-timeout:
 			// the worker fuck up, release the task
 			log.Error("Worker: ", workerID, "fuck up")
-			m.scheduler.ReportTask(m.workerTaskMap[workerID], ResultFailed)
```
Comment (Lcch) [inline master/master.go]: Keep those functions here. It is able to lock master when is needed.

File: master/master.go [hunk 3]
```diff
@@ -157,18 +162,26 @@ func (s *server) RequestTask(ctx context.Context, in *kmrpb.RegisterParams) (*km
...
+	var ok bool
+	if t, ok = s.master.workerTaskMap[in.WorkerID]; !ok {
 		log.Errorf("WorkerID %v is not working on anything", in.WorkerID)
 		return &kmrpb.Response{Retcode: 0}, nil
 	}
 
 	var heartbeatCode int
 	switch in.Retcode {
 	case kmrpb.ReportInfo_FINISH:
+		s.master.scheduler.ReportTask(t, ResultOK)
```
Comment (Lcch) [inline master/master.go]: keep it at HearBeat handler

File: master/scheduler.go [hunk 4]
```diff
@@ -208,61 +223,76 @@ func (s *Scheduler) StartSchedule(visitor EventHandler) error {
...
+							break
 						}
 					}
+					if finished {
+						break
+					}
 				}
-				c <- requestJobOutput{TaskDescription{}, nil}
+				c <- out
+				close(c)
```
Comment (Lcch) [inline master/scheduler.go]: Close channel at function who delivers requestJobChan

File: master/scheduler.go [hunk 5]
```diff
@@ -176,16 +177,30 @@ func (s *Scheduler) StartSchedule(visitor EventHandler) error {
...
+				}
+				if _, ok := s.reducerFinishedCnt[j]; ok {
+					log.Panic("Job repeated scheduled, reducerCnt != 0")
+				}
+				if _, ok := s.phaseMap[j]; ok {
+					log.Panic("Job repeated scheduled, phase != nil")
+				}
+				s.mapperFinishedCnt[j], s.reducerFinishedCnt[j], s.phaseMap[j] = 0, 0, mapPhase
 				c.output <- state
+				close(c.output)
```
Comment (Lcch) [inline master/scheduler.go]: same as below

File: util/k8sutil.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
+package util
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (Lcch) [inline util/k8sutil.go]: import style

File: master/master.go (hunk 3 above)
Comment (KKRainbow) [inline master/master.go]: done

File: cli/entry.go (hunk 1 above)
Comment (KKRainbow) [inline cli/entry.go]: done

File: master/master.go (hunk 2 above)
Comment (KKRainbow) [inline master/master.go]: done

File: util/k8sutil.go (hunk 6 above)
Comment (KKRainbow) [inline util/k8sutil.go]: done

File: examples/charactercount/example.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+	"github.com/naturali/kmr/jobgraph"
+	"github.com/naturali/kmr/mapred"
+	"github.com/naturali/kmr/cli"
+)
+
+const (
+	maxWordLength = 20
+)
+
+type wordCountMap struct {
```
Comment (Lcch) [inline examples/charactercount/example.go]: characterCountMap

File: executor/executor.go [hunk 8]
```diff
@@ -37,32 +37,34 @@ func (cw *ComputeWrapClass) BindReducer(reducer mapred.Reducer) {
 
 func (cw *ComputeWrapClass) BindCombiner(combiner mapred.Reducer) {
 	// XXX: should use reducer directly
```
Comment (Lcch) [inline executor/executor.go]: // TODO:

File: executor/worker.go [hunk 9]
```diff
@@ -30,19 +33,32 @@ type Worker struct {
...
 		flushOutSize: flushOutSize,
 	}
 	return &worker
 }
 
+func (w *Worker) getBucket(files jobgraph.Files) bucket.Bucket {
+	var bk bucket.Bucket
+	switch files.GetBucketType() {
+	case jobgraph.MapBucket:
+		bk = w.mapBucket
```
Comment (Lcch) [inline executor/worker.go]: return

**Pull request merged**